### PR TITLE
Only rehash if a write may have failed

### PR DESCRIPTION
**Pull request by mkeeter**
Previously, we always rehashed data during a `flush` operation, which required reading it back from disk and computing an integrity hash.

However, if a block was successfully written, then we already know what its hash value will be (since we computed it earlier)!  We _only_ need to read + rehash if we failed to write the data, because if we failed to write the data, then the on-disk bytes are in an unknown state.

This PR tracks dirty blocks as a map from `block -> Option<u64>`, so that if the hash is known, it won't be recomputed during a flush operation.

It shows a small but noticeable (1-3%) performance increase when running `pgbench`:

Screenshot 2023-08-29 at 11 48 39 AM

### CODE REVIEW

File: downstairs/src/region.rs [hunk 1]
```diff
@@ -1285,6 +1316,20 @@ impl Extent {
         // Write any remaining data
         batched_pwritev.perform_writes()?;
 
+        // At this point, we know that the on-disk values for any written blocks
+        // must match the integrity hashes calculated above (and stored to
+        // SQLite).  We can therefore store hash values for many dirty blocks,
```
Comment (jmpesp) [inline downstairs/src/region.rs:1321]: Unfortunately, we don't know that until `flush` returns Ok - `write` returning ok is insufficient to know that bytes have landed in durable storage.

**Review by jmpesp — CHANGES_REQUESTED** [review]

File: downstairs/src/region.rs (hunk 1 above)
Comment (jmpesp) [reply downstairs/src/region.rs:1321]: After a few re-reads of this PR, this comment needs to change but at this point I think the PR itself is sound.

Comment (mkeeter) [reply downstairs/src/region.rs]: Yeah, I was struggling for the right way to phrase "written-but-not-flushed" data, but "on-disk" is definitely misleading.

Changed in 485307332c91e99af6f35abf169475c6fb997954, let me know what you think of the new working.

Comment (jmpesp) [reply downstairs/src/region.rs]: :+1:

**Review by faithanalog — APPROVED** [review]

File: downstairs/src/region.rs [hunk 2]
```diff
@@ -1203,9 +1232,10 @@ impl Extent {
             (job_id, self.number, writes.len() as u64)
         });
 
+        let mut hashes_to_write = vec![];
```
[the remark this answers is not in the document]
Comment (mkeeter) [reply downstairs/src/region.rs:1235]: yeah, seems like best practice (although it probably won't be noticeable in the performance test)

**Pull request merged**